gap/attributes/isomorph.gi

```diff
@@ -171,50 +190,150 @@ end);
 InstallMethod(IsomorphismSemigroups, "for finite monogenic semigroups",
 [IsMonogenicSemigroup and IsFinite, IsMonogenicSemigroup and IsFinite],
 function(S, T)
-  local genS, genT, SS, TT;
+  local s, t, SS, TT, map, inv;
+
+  if IsSimpleSemigroup(S) then
+    if IsSimpleSemigroup(T) then
+      return IsomorphismSemigroups(S, T);
+    else
+      return fail;
+    fi;
+  elif Size(S) <> Size(T) or NrDClasses(S) <> NrDClasses(T) then
+    return fail;
+  fi;
 
   # Monogenic semigroups are of the same size, are not groups, and have the
-  # same number of D-classes by this point, and so they are isomorphism
-  genS := Representative(MaximalDClasses(S)[1]);
-  genT := Representative(MaximalDClasses(T)[1]);
-  SS := Semigroup(genS);
-  TT := Semigroup(genT);
-  return MagmaIsomorphismByFunctionsNC(S, T,
-           x -> genT ^ Length(Factorization(SS, x)),
-           x -> genS ^ Length(Factorization(TT, x)));
+  # same number of D-classes by this point, and so they are isomorphic
+  s := Representative(MaximalDClasses(S)[1]);
+  t := Representative(MaximalDClasses(T)[1]);
+  SS := Semigroup(s);
+  TT := Semigroup(t);
+  map := x -> t ^ Length(Factorization(SS, x));
+  inv := x -> s ^ Length(Factorization(TT, x));
+  return MagmaIsomorphismByFunctionsNC(S, T, map, inv);
 end);
 
+SEMIGROUPS.IsomorphismInvariantDigraph := function(S)
+  local M, n, Color1Node, Color2Node, Color3Node, Widget, out, colors, x, y, z;
+
+  M := MultiplicationTable(S);
+  n := Size(S);
+
+  # The original nodes
+  Color1Node := function(i)
+    return i;
+  end;
+
+  # i = 1 .. n, j = 1 .. n
+  Color2Node := function(i, j)
+    Assert(1, 1 <= i and i <= n);
+    Assert(1, 1 <= j and j <= n);
+    return i * n + j;
+  end;
+
+  Color3Node := function(i, j)
```

Comment:
I haven't worked out why these nodes need to exist yet.

Reply:
These ones actually don’t need to exist but in some experiments their presence improved the performance so I left them in.

Reply:
These ones actually don’t need to exist but in some experiments their presence improved the performance so I left them in.

Reply:
Fair enough!

Reply:
Probably this falls under the heading “the graph encoding can probably be improved”

Reply:
I actually removed the Color3Nodes now, same performance, or slightly better